### PR TITLE
[MGDOBR-1383] - Provision a KafkaSource in the reconcile loop of the …

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
…ManagedBridge

Jira : https://issues.redhat.com/browse/MGDOBR-1383

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-XXX] - $clear_explanation_of_what_you_did"
- [ ] The layers in the `kustomize` folder have been updated accordingly.
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: shard-operator-parent/shard-operator-v2/src/main/resources/templates/kafka-source.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  consumerGroup: knative-group
+  bootstrapServers:
+    - 'rhose-loca-c--o--mmk-hbcp-irvga.bf2.kafka.rhcloud.com:443'
+  topics:
+    - 'vajain_rhoc_topic'
+  sink:
+    ref:
+      apiVersion: eventing.knative.dev/v1
+      kind: Broker
+      name: rhose-topic-broker
```
Comment (r00ta) [inline shard-operator-parent/shard-operator-v2/src/main/resources/templates/kafka-source.yaml:32]: you might want to use better placeholders in this template

**Review by r00ta — COMMENTED** [review]
Hi @vaibhavjainwiz , thanks for the PR. But as per chat given that v2 has been stopped entirely it does not make much sense anymore to implement these features.

File: shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatus.java [hunk 2]
```diff
@@ -63,6 +63,9 @@ public void markConditionFalse(final String conditionType, final String reason,
             condition.get().setStatus(ConditionStatus.False);
             condition.get().setLastTransitionTime(new Date());
             this.conditions.add(condition.get());
+        } else {
```
Comment (manstis) [inline shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatus.java:66]: I'm against this personally.

**ALL** `Condition`'s needed to create a resource **MUST** be present from the beginning. The `manager` uses the whole set to determine if a resource is `READY` and adding some in "later" risks the interpretation being wrong if performed before _other_ `Condition`'s are added.

File: shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatus.java [hunk 3]
```diff
@@ -82,6 +85,9 @@ public void markConditionTrue(final String conditionType, final String reason) {
             condition.get().setStatus(ConditionStatus.True);
             condition.get().setLastTransitionTime(new Date());
             this.conditions.add(condition.get());
+        } else {
```
Comment (manstis) [inline shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatus.java:88]: Likewise.

File: shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatus.java [hunk 4]
```diff
@@ -116,4 +122,8 @@ public void markConditionFailed(final String conditionType) {
         markConditionFailed(conditionType, null, "", null);
     }
 
+    public void removeCondition(String conditionType) {
```
Comment (manstis) [inline shard-operator-parent/shard-operator-core/src/main/java/com/redhat/service/smartevents/shard/operator/core/resources/CustomResourceStatus.java:125]: I don't think we should EVER remove a `Condition`.

If, for whatever reason, the `Condition` is already `True` (IDK, perhaps a shared Kafka Source/Sink?) then the `Condition` should simply be marked as `READY`.

File: shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedBridgeServiceImpl.java [hunk 5]
```diff
@@ -97,34 +110,86 @@ private void ensureManagedBridgeLabels(ManagedBridge managedBridge, final HasMet
...
 
-        KafkaConnectionDTO kafkaConnection = bridgeDTO.getKnativeBrokerConfiguration().getKafkaConnection();
-        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_BOOTSTRAP_SERVERS_SECRET,
-                Base64.getEncoder().encodeToString(kafkaConnection.getBootstrapServers().getBytes()));
-        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_USER_SECRET, Base64.getEncoder().encodeToString(kafkaConnection.getClientId().getBytes()));
-        expected.getData().put(GlobalConfigurationsConstants.KNATIVE_KAFKA_PASSWORD_SECRET, Base64.getEncoder().encodeToString(kafkaConnection.getClientSecret().getBytes()));
+        Secret expected = templateProvider.loadBridgeIngressSecretTemplate(managedBridge, new TemplateImportConfig(V2_OPERATOR_NAME)
+                .withNamespaceFromParent()
+                .withOwnerReferencesFromParent());
+        expected.getMetadata().setName(KNATIVE_BROKER_SECRET_NAME);
```
Comment (manstis) [inline shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedBridgeServiceImpl.java:118]: Why can this not be done in the `loadBridgeIngressSecretTemplate(...)` method? It seems strange that the name is set in the metadata _outside_ of the `TemplateProvider` that updates all other metadata. I'd rather `name` either be added to `TemplateImportConfig` (and not just the ability to set it from parent) or overload the `loadBridgeIngressSecretTemplate(...)` method.

File: shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedBridgeServiceImpl.java [hunk 6]
```diff
@@ -97,34 +110,86 @@ private void ensureManagedBridgeLabels(ManagedBridge managedBridge, final HasMet
...
+            LOGGER.info("Create/Update Secret with name '{}' for ManagedBridge with id '{}' in namespace '{}'.", expected.getMetadata().getName(), managedBridge.getMetadata().getName(),
+                    managedBridge.getMetadata().getNamespace());
+        }
+    }
+
+    private void createOrUpdateDNSConfigurationSecret(ManagedBridge managedBridge, DNSConfigurationDTO dnsConfiguration) {
+        Secret expected = templateProvider.loadBridgeIngressSecretTemplate(managedBridge, new TemplateImportConfig(V2_OPERATOR_NAME)
+                .withNamespaceFromParent()
+                .withOwnerReferencesFromParent());
+        expected.getMetadata().setName(DNS_CONFIGURATION_SECRET_NAME);
```
Comment (manstis) [inline shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedBridgeServiceImpl.java:147]: Likewise.

File: shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedBridgeServiceImpl.java [hunk 7]
```diff
@@ -97,34 +110,86 @@ private void ensureManagedBridgeLabels(ManagedBridge managedBridge, final HasMet
...
+                    managedBridge.getMetadata().getNamespace());
+        }
+    }
+
+    private void createOrUpdateSourceConfigurationSecret(ManagedBridge managedBridge, SourceConfigurationDTO sourceConfiguration) {
+
+        Secret expected = templateProvider.loadBridgeIngressSecretTemplate(managedBridge, new TemplateImportConfig(V2_OPERATOR_NAME)
+                .withNamespaceFromParent()
+                .withOwnerReferencesFromParent());
+        expected.getMetadata().setName(SOURCE_CONFIGURATION_SECRET_NAME);
```
Comment (manstis) [inline shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/ManagedBridgeServiceImpl.java:174]: Likewise.

File: shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/controllers/ManagedBridgeController.java [hunk 8]
```diff
@@ -125,6 +131,22 @@ public UpdateControl<ManagedBridge> reconcile(ManagedBridge managedBridge, Conte
             status.markConditionTrue(DP_NETWORK_RESOURCE_READY_NAME);
         }
 
+        // Create/Update KafkaSource
```
Comment (manstis) [inline shard-operator-parent/shard-operator-v2/src/main/java/com/redhat/service/smartevents/shard/operator/v2/controllers/ManagedBridgeController.java:134]: I'm not a fan of the `Condition`'s being added to/removed from in the `Reconciler`.

Can we not determine if a Kafka Source is needed from the incoming payload (in `ManagedBridgeService.createManagedBridge(..)` and initialise the set of `Conditions` correctly there?

**Review by manstis — COMMENTED** [review]
FWIW I have some comments.

Are there tests missing too? I don't see where `KafkaSourceServiceImpl` is tested for example.

I also agree with @r00ta comment... Should we just close all PRs.

**Comment by manstis** [comment]
It was proposed in today's stand-up that all `v2` PRs be closed.

`v2` is effectively dead.

**Pull request closed without merging**